WIP: add map revealing all airports and tall obstacles

db48x opened the pull request:
#### Summary
Features "Add a new type of map that reveals all airports and tall obstacles in the world"

#### Purpose of change

Real–world pilots use Aeronautical Section Charts to plot cross–country routes. They include many relevant details about every airport in the country, including all the thousands of smaller general avaition airports. Tall obstacles such as wind turbines and tall buildings are marked, as are sources of bright light such as stadiums. Urban areas, interstate highways, and major rail routes are also marked for improved situational awareness. All navigational beacons are plotted, and all control towers have their radio frequencies indicated. Airspace limitations and restrictions are prominant.

#### Describe the solution

This commit adds a new type of map to the game, tagged with WORLD_MAP. This new type of map works a little differently from the existing maps. The area revealed is centered around the midlands, rather than around the character who reads them. Instead of being labeled with a city name, they are labled with the name of the game world. Instead of finding multiple maps and reading each one once, the player is only allowed to read each type of world map once; multiple world maps convey no additional information.

The air chart currently implemented reveals terrain in a square 24 overmaps on a side, centered on the midlands. It takes my computer several minutes to generate all of that terrain, and several gigabytes of memory. During that time there is no feedback to the player that a long–running task is taking place. It’s a pretty bad experience, actually. Use with caution.

#### Describe alternatives you've considered

What we really should do is rearrange the terrain generation so that the locations of certain items within an overmap can be determined without needing to completely generate everything in the overmap.

Do we spawn items and monsters in these overmaps, even if they are far from the player? If so, perhaps the simplest change would be to defer that work and only generate the overmap terrain without actually doing the rest of the work. I don’t know enough about how it works, however, to make concrete recommendations.

#### Testing

Just manual testing.

#### Additional context

I picked the radius so that it would reveal the coastline, which is definitely included on the real maps.

Fixes #70327

db48x committed:
add map revealing all airports and tall obstacles

Real–world pilots use Aeronautical Section Charts to plot
cross–country routes. They include many relevant details about every
airport in the country, including all the thousands of smaller general
avaition airports. Tall obstacles such as wind turbines and tall
buildings are marked, as are sources of bright light such as
stadiums. Urban areas, interstate highways, and major rail routes are
also marked for improved situational awareness. All navigational
beacons are plotted, and all control towers have their radio
frequencies indicated. Airspace limitations and restrictions are
prominant.

This commit adds a new type of map to the game, tagged with
WORLD_MAP. This new type of map works a little differently from the
existing maps. The area revealed is centered around the midlands,
rather than around the character who reads them. Instead of being
labeled with a city name, they are labled with the name of the game
world. Instead of finding multiple maps and reading each one once, the
player is only allowed to read each type of world map once; multiple
world maps convey no additional information.

The air chart currently implemented reveals terrain in a square 24
overmaps on a side, centered on the midlands. It takes my computer
several minutes to generate all of that terrain, and several gigabytes
of memory. During that time there is no feedback to the player that a
long–running task is taking place. It’s a pretty bad experience,
actually. Use with caution.

Fixes #70327

John-Candlebury commented:
Makes sense to add this to the helicopter from the "Last Flight" scenario

db48x commented:
That’s a great point; if you were just flying then you must have your maps ready to hand. And the road map it has doesn’t make much sense to me. However, you do have to be a helicopter pilot in that scenario, and I added one to their starting inventory, so that’s already taken care of.

However, I suppose I could add a small chance of a map to every other helicopter in the game. They must get left behind occasionally, right?

Inglonias commented:
Depends who the pilot is, the circumstances of the crash, and what the map is made of (if it's just paper, odds are good it would be destroyed in a fiery or forceful crash)

db48x commented:
The Last Flight scenario doesn’t have a wreck. You’re a pilot who landed their helicopter when they ran out of fuel.

db48x committed:
add a small chance of an air chart in every plane

Pilots are supposed to carry them with them, but they must get left
behind occasionally. You might find one even in a wreck, unless the
whole thing is a burnt–out shell.

db48x committed:
placate style checker

Two spaces, really? We’re not using typewriters here.

Procyonae commented:
>The air chart currently implemented reveals terrain in a square 24 overmaps on a side, centered on the midlands. It takes my computer several minutes to generate all of that terrain, and several gigabytes of memory. During that time there is no feedback to the player that a long–running task is taking place. It’s a pretty bad experience, actually. Use with caution.

That's not reasonable

>Do we spawn items and monsters in these overmaps, even if they are far from the player?

No only those in the (comparatively tiny and already loaded) reality bubble

db48x commented:
> > Do we spawn items and monsters in these overmaps, even if they are far from the player?
> 
> No only those in the (comparatively tiny and already loaded) reality bubble

Well, so much for that brilliant idea :)

I did come up with a different way to speed up the process by around 50%, but that’s still not quick enough. I also added a test, but got stuck on how to verify that the map generation actually worked. (The goal of the test is primarily to have a performance test–case that doesn’t require any keyboard input.) I wonder if I should just timebox the test and revisit it later, or if an accurate idea of just how long reading this map takes is useful.

#70404 could also be useful for speeding it up.

prharvey commented:
> > > Do we spawn items and monsters in these overmaps, even if they are far from the player?
> > 
> > 
> > No only those in the (comparatively tiny and already loaded) reality bubble
> 
> Well, so much for that brilliant idea :)
> 
> I did come up with a different way to speed up the process by around 50%, but that’s still not quick enough. I also added a test, but got stuck on how to verify that the map generation actually worked. (The goal of the test is primarily to have a performance test–case that doesn’t require any keyboard input.) I wonder if I should just timebox the test and revisit it later, or if an accurate idea of just how long reading this map takes is useful.
> 
> #70404 could also be useful for speeding it up.

Micro-optimizations like what you linked won't make a big difference in this case. Optimizing something that runs for several minutes down to a few seconds would need an entirely new approach.

db48x commented:
> Micro-optimizations like what you linked won't make a big difference in this case. Optimizing something that runs for several minutes down to a few seconds would need an entirely new approach.

Don’t knock them. The only way software ever achieves real efficiency is by having developers spend years chasing every single 1% win. Individually each one is small, but over time they really add up. Plus, they’re bound to trip over a few 10% or 20% wins along the way. But you’re right, for the purposes of this bug it doesn’t change the fact that I need to add a progress indicator so that the user doesn’t think the game has hung. (And ultimately we need to change the map generation so that high–level questions of this type can be answered quickly, without invoking the full map generation.) I never intended to imply that they would.

On a re–review of the profiles I have collected, the `string_id`/`int_id` changes mentioned in #70399 may be a bigger win during map generation. On the other hand, my own little improvement to how we search the map for things to reveal slashes a whole bunch of calls into both the `string_id` code and the `divide_xy_round_to_minus_infinity` stuff, so both of them become much smaller improvements. The profile is now 97% pure map generation; the search barely shows up around the edges.

Speaking of map generation, would maps still be generated correctly if we skipped `overmap::place_roads` until the player actually approaches the overmap? That’s 15% right there. What about if we skipped `overmap::place_forest_trails` (4%), and `overmap::place_forest_trailheads` (12%). The airports are never inside forests, so delaying those shouldn’t hurt, right? I think I am going to push such questions into their own issue though.

TealcOneill commented:
Other than the failing check was this finished? This seems like a great feature and was something I was thinking of doing but you've done a much better job than I would have.

Maleclypse commented:
Conflicts to resolve?

db48x commented:
I have no plans for revisiting this idea until after work on #70056 was begun. Once we have some kind of high–level regional map, I would add an international airport to New Bristol, overlapping with the coast, smaller airports to Bunyan and Montauk, and then a decent chance of a small or private airport to every other region on the map.

Then instead of searching every OMT for airport features like runways and towers, I would make a special map type that directly examines the regional map instead. This would prevent the game from taking ages to fully generate the map of the whole state just to show a dozen airports.

Night-Pryanik commented:
@db48x so it's ok to close then?

Night-Pryanik commented:
Closing as stale. If you wish to continue working on this PR, ping me to reopen.
